pkg/util/vsphere/vm/hardware_version_test.go

```diff
@@ -0,0 +1,225 @@
+			expectedErr                   error
+			expectedResult                bool
+			fetchProperties               bool
+			initialHardwareVersion        string
+			minHardwareVersion            int32
+			expectedHardwareVersion       string
+			skipPostReconcileVerification bool
+		)
+
+		BeforeEach(func() {
+			ctx = suite.NewTestContextForVCSim(builder.VCSimTestConfig{})
+			vmList, err := ctx.Finder.VirtualMachineList(ctx, "*")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(vmList).ToNot(BeEmpty())
+			mgdObj = vmutil.ManagedObjectFromMoRef(vmList[0].Reference())
+			simulator.TaskDelay.MethodDelay = map[string]int{}
+			obj = object.NewVirtualMachine(ctx.VCClient.Client, mgdObj.Self)
+			expectedResult = false                // default
+			expectedErr = nil                     // default
+			expectedHardwareVersion = ""          // default
+			fetchProperties = true                // default
+			minHardwareVersion = 0                // default
+			skipPostReconcileVerification = false // default
+		})
+
+		JustBeforeEach(func() {
+			if obj != nil && initialHardwareVersion != "" {
+				t, err := obj.Reconfigure(
+					ctx,
+					types.VirtualMachineConfigSpec{
+						Version: initialHardwareVersion,
+					})
+				Expect(err).ToNot(HaveOccurred())
+				ti, err := t.WaitForResultEx(ctx)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(ti.Error).To(BeNil())
+			}
+
+			// The actual test
+			result, err := vmutil.ReconcileMinHardwareVersion(
+				logr.NewContext(ctx, suite.GetLogger()),
+				ctx.VCClient.Client,
+				mgdObj,
+				fetchProperties,
+				minHardwareVersion)
+
+			e, a := expectedErr, err
+			switch {
+			case e != nil && a != nil:
+				Expect(a.Error()).To(Equal(e.Error()))
+			case e == nil && a != nil:
+				Fail("unexpected error occurred: " + a.Error())
+			case e != nil && a == nil:
+				Fail("expected error did not occur: " + e.Error())
+			case e == nil && a == nil:
+				Expect(err).ToNot(HaveOccurred())
+				Expect(result).To(Equal(expectedResult))
+				if !skipPostReconcileVerification {
+					Expect(obj.Properties(
+						ctx,
+						mgdObj.Reference(),
+						[]string{vmutil.HardwareVersionProperty},
+						&mgdObj),
+					).To(Succeed())
+					Expect(mgdObj.Summary.Config.HwVersion).To(Equal(
+						expectedHardwareVersion))
+				}
+			}
+		})
+
+		AfterEach(func() {
+			ctx.AfterEach()
+			ctx = nil
+		})
+
+		When("The VM does not exist", func() {
+			BeforeEach(func() {
+				minHardwareVersion = 15
+				mgdObj.Self.Value = "does-not-exist"
+				initialHardwareVersion = "vmx-13"
+				expectedErr = errors.New("failed to retrieve properties ServerFaultCode: The object has already been deleted or has not been completely created")
+			})
+			Context("and the hardware version is cached in the managed object", func() {
+				BeforeEach(func() {
+					mgdObj.Summary.Config.HwVersion = "vmx-13"
+				})
+				Context("and fetchProperties is true", func() {
+					It("an upgrade op should return an error", func() {})
```

Comment:
Can we please move the validations in the respective `It` blocks?  While it might add a few additional lines of code, think that will help the readability massively.

For example, we could have 
```golang
			result, err := vmutil.ReconcileMinHardwareVersion(
				logr.NewContext(ctx, suite.GetLogger()),
				ctx.VCClient.Client,
				mgdObj,
				fetchProperties,
				minHardwareVersion)
```

in the `JustBeforeEach` block and we verify the `result` and the `err` here.

Reply:
Actually, you couldn't have it as you wrote it as it would scope `result` and `err` to the `JustBeforeEach` block :)

As for the suggestion, I need to give it more thought. My initial take is that this is part and parcel of Ginkgo DSL, and that folks should embrace it. But like I said, I need to marinate on it. Either way I will not change the tests in this PR as they follow the same pattern from `power_state_test.go`, and I do not want to rewrite both right now. 

And besides, this is kind of the purpose of Ginkgo -- optimizing the tests.

Reply:
I addressed this in the latest rebase.